File: solutions/professional_2023/friends_walks.cpp
#include <iostream>
#include <vector>
using namespace std;

const long long MOD = 1000000007;

void count_values_SLOW(vector<long long>& values, int k) {
    int n = values.size();
    long long overall_sum = 0;

    for (int i = 0; i < n; ++i) {
        overall_sum += values[i];
        overall_sum %= MOD;
    }

    for (int step = 0; step < k; ++step) {

        for (int i = 0; i < n; ++i) {
            long long new_coord = (overall_sum - values[i]) % MOD;
            overall_sum = (overall_sum - values[i]) % MOD + new_coord;
            overall_sum %= MOD;

            values[i] = new_coord;
        }
    }

}

int main() {
    int t;
    cin >> t;

    while (t--) {
        int n, k;
        cin >> n >> k;
        vector<long long> coords(n);

        for (int i = 0; i < n; ++i) {
            cin >> coords[i];
        }

        vector<long long> final_positions = count_values(coords, k);

        for (int i = 0; i < n; ++i) {
            cout << final_positions[i] << " ";
        }
        cout << endl;
    }

    return 0;
}

File: solutions/professional_2023/friends_walks.py
MOD = 1000000007


def count_values(initial_values: list[int], k):
    overall_sum = sum(initial_values) % MOD
    
    n = len(initial_values)
    
    
    for step in range(k):
        
        for i in range(n):
            coord = initial_values[i]
            
            overall_sum -= coord
            initial_values[i] = overall_sum

            overall_sum += overall_sum
            overall_sum %= MOD

    return initial_values
            
t = int(input())

for _ in range(t):
    
    n, k = map(int, input().split())
    
    coords = list(map(int, input().split()))
    print(*count_values(coords, k))